### PR TITLE
[B] Added Location.to{World,Local}{Axis,}(Vector). Adds BUKKIT-5338

### PR DESCRIPTION
**Pull request by TomyLobo**
#### The Issue

The Location class currently lacks methods for transforming between coordinate systems easily.
#### Justification for this PR

This PR allows developers to do the affine transformation to and from a local coordinate system that rotates and moves with the reference Location, whereby new Location(*, 0, 0, 0, 0, 0) does no transformation at all.

This implies that:
- Positive y is up
- Positive z (south, i.e. yaw=0) is forward
- Positive x (east, which is left of south) is left
- getDirection() is equivalent to toWorldAxis(new Vector(0,0,1))

This can be used to render effects adjacent to the player or creature and to spawn projectiles in a consistent pattern relative to the player's view.
#### PR Breakdown

The toWorld\* methods transform from local coordinates to world coordinates.
The toLocal\* methods transform from world coordinates to local coordinates.
The *Axis methods deal with axes instead of positions.
#### Testing Results and Materials

Unit tests are included and should be sufficient to ensure that the new methods are working.
#### Relevant PR(s)

[B-1012](https://github.com/Bukkit/Bukkit/issues/1012)
#### JIRA Ticket

[BUKKIT-5338](https://bukkit.atlassian.net/browse/BUKKIT-5338)

### CODE REVIEW

**Comment by turt2live** [comment]
I can honestly say I have no idea what the math involved here is doing/should do, but I'll look over the formatting in a moment.

In the meantime, please update your title to match the CONTRIBUTING.md guidelines.

File: src/main/java/org/bukkit/Location.java [hunk 1]
```diff
@@ -557,4 +557,106 @@ public Location clone() {
     public static int locToBlock(double loc) {
         return NumberConversions.floor(loc);
     }
+
+    /**
+     * Transforms an axis from the local reference frame described by this
+     * location into the world reference frame.
+     * <p>
+     * x=left, y=up, z=forward
```
Comment (turt2live) [inline src/main/java/org/bukkit/Location.java]: These 3 variables should be in sentence form or list form with an applicable sentence. 

Such as `.. where x represents the left...`

or `... where the following conditions are applied: <list>`

Comment (TomyLobo) [inline src/main/java/org/bukkit/Location.java]: I'll use the later. thanks for the feedback.

File: src/main/java/org/bukkit/Location.java [hunk 2]
```diff
@@ -557,4 +557,106 @@ public Location clone() {
...
+    public Vector toWorldAxis(Vector axis) {
+        final double yaw = Math.toRadians(this.getYaw());
+        final double pitch = Math.toRadians(this.getPitch());
+
+        final double cosYaw = Math.cos(yaw);
+        final double sinYaw = Math.sin(yaw);
+        final double cosPitch = Math.cos(pitch);
+        final double sinPitch = Math.sin(pitch);
+
+        final Vector forward = new Vector(
```
Comment (turt2live) [inline src/main/java/org/bukkit/Location.java]: This does not need to be spread over 5 lines, nor do the following vectors.

Additionally, `x * x` should be used as a format, not `x*x` (note the spacing)

File: src/main/java/org/bukkit/Location.java [hunk 3]
```diff
@@ -557,4 +557,106 @@ public Location clone() {
...
+    public Vector toLocalAxis(Vector axis) {
+        final double yaw = Math.toRadians(this.getYaw());
+        final double pitch = Math.toRadians(this.getPitch());
+
+        final double cosYaw = Math.cos(yaw);
+        final double sinYaw = Math.sin(yaw);
+        final double cosPitch = Math.cos(pitch);
+        final double sinPitch = Math.sin(pitch);
+
+        final Vector xAxis = new Vector(
```
Comment (turt2live) [inline src/main/java/org/bukkit/Location.java:628]: https://github.com/Bukkit/Bukkit/pull/1012/files#r9171780

File: src/main/java/org/bukkit/Location.java [hunk 4]
```diff
@@ -557,4 +557,106 @@ public Location clone() {
...
+        );
+
+        return left.multiply(axis.getX()).add(up.multiply(axis.getY())).add(forward.multiply(axis.getZ()));
+    }
+
+    /**
+     * Transforms an axis from the world reference frame into the local reference frame described by this
+     * location.
+     * <p>
+     * x=left, y=up, z=forward
```
Comment (turt2live) [inline src/main/java/org/bukkit/Location.java]: https://github.com/Bukkit/Bukkit/pull/1012/files#r9171772

File: src/main/java/org/bukkit/Location.java [hunk 5]
```diff
@@ -557,4 +557,106 @@ public Location clone() {
...
+        );
+
+        return xAxis.multiply(axis.getX()).add(yAxis.multiply(axis.getY())).add(zAxis.multiply(axis.getZ()));
+    }
+
+    /**
+     * Transforms a position from the local reference frame described by this
+     * location into the world reference frame.
+     * <p>
+     * x=left, y=up, z=forward
```
Comment (turt2live) [inline src/main/java/org/bukkit/Location.java]: https://github.com/Bukkit/Bukkit/pull/1012/files#r9171772

File: src/main/java/org/bukkit/Location.java (hunk 2 above)
Comment (TomyLobo) [inline src/main/java/org/bukkit/Location.java]: fixed locally

File: src/main/java/org/bukkit/Location.java (hunk 3 above)
Comment (TomyLobo) [inline src/main/java/org/bukkit/Location.java:628]: fixed locally

File: src/main/java/org/bukkit/Location.java [hunk 6]
```diff
@@ -557,4 +557,106 @@ public Location clone() {
...
+     */
+    public Vector toWorld(Vector position) {
+        return toWorldAxis(position).add(toVector());
+    }
+
+    /**
+     * Transforms a position from the world reference frame into the local reference frame described by this
+     * location.
+     * <p>
+     * x=left, y=up, z=forward
```
Comment (turt2live) [inline src/main/java/org/bukkit/Location.java]: https://github.com/Bukkit/Bukkit/pull/1012/files#r9171772

**Comment by TomyLobo** [comment]
I pushed changes addressing your comments.
the maths:
Basically this is 2 parametric rotation matrices (one for yaw, one for pitch), multiplied into one in a specific order.
in toWorldAxis (582-598), you see the matrix split up into 3 vectors.
Since rotation matrices are orthonormal, you can invert them by simply transposing them.
You see that inverted matrix in toLocalAxis (624-640), again split up into 3 vectors.

Each of these 3 vectors is one of the base axes of the target coordinate system, represented in the source coordinate system.
Multiplying them by the components corresponding to those axes and adding it all up yields the transformed axis.

File: src/main/java/org/bukkit/Location.java [hunk 7]
```diff
@@ -557,4 +557,120 @@ public Location clone() {
     public static int locToBlock(double loc) {
         return NumberConversions.floor(loc);
     }
+
+    /**
+     * Transforms an axis from the local reference frame described by this
+     * location into the world reference frame.
+     * <p>
+     * The following conditions apply:
+     * <li>X axis points to the left</li>
```
Comment (Wolvereness) [inline src/main/java/org/bukkit/Location.java]: We don't use closing list item markers, and this needs a unordered list market before it starts.

**Comment by Wolvereness** [comment]
I normally take pride in my knowledge of mathematical concepts, but I have no idea what this is supposed to do. Intended input and output, with use case isn't documented very well. Based on what little description you gave, this is already possible using the current api, without hand-rolling the math.

I also want to see unit tests parameterized with expected inputs and outputs, like https://github.com/Bukkit/Bukkit/blob/master/src/test/java/org/bukkit/LocationTest.java

Oh, and if you have to assign an arbitrary direction as 'forward' , I'm pretty sure it's not appropriate api.

**Comment by TomyLobo** [comment]
> I normally take pride in my knowledge of mathematical concepts, but I have no idea what this is supposed to do.

I explained it in [my comment](https://github.com/Bukkit/Bukkit/pull/1012#issuecomment-33291571).

> Intended input and output, with use case isn't documented very well.

Described in the justification and by the test cases.
I'll translate the test cases into plain english and attach them to the javadocs.

> Based on what little description you gave, this is already possible using the current api, without hand-rolling the math.

Can you explain how?
EDIT: I did a "find usages" on Math.cos and found nothing outside Location.java and only getDirection inside Location.java.
And seeing as a cosine would be necessary for this, I'm going to disagree here and claim that there is _no_ functionality like this in Bukkit yet.

> I also want to see unit tests parameterized with expected inputs and outputs, like https://github.com/Bukkit/Bukkit/blob/master/src/test/java/org/bukkit/LocationTest.java

I don't know how these parameterized tests work, but I did see that test you made. I roughly see what's going on, but seeing as I had all of my tests parametrized already, I don't see how it's necessary to make them conform to that framework you're using that I'm not familiar with.

> Oh, and if you have to assign an arbitrary direction as 'forward' , I'm pretty sure it's not appropriate api.

I agree with that statement.
However, if you face direction 0 (south) and then walk forward, you'll notice z goes up, so assigning positive z to forward isn't arbitrary.
This kind of conformity is covered by a unit test, too. I'd be happy to explain these to you if you ping me on IRC.
The gist of that conversation can then become a comment or javadoc or something.

_(edited the last time at 10:08 CET, 16 minutes after posting)_

**Comment by TomyLobo** [comment]
thought of a better solution

**Comment by Wolvereness** [comment]
Context: https://gist.github.com/Wolvereness/8631025

**Comment by TomyLobo** [comment]
Status: After messing around with the matrix multiplication order and angle units for a bit, I made it conform to the tests I used for the previous proposal.
That implies, however, that I'm not testing roll at all and there's nothing in the API that uses roll which I could ensure conformity to with another test.
That test framework where you specify inputs and outputs would be useful here :)

**Comment by Wolvereness** [comment]
What happened to this?

**Comment by TomyLobo** [comment]
On #bukkitdev, I defended a guy who got banned for using "stuff.com" (a malware site) as a placeholder.
I shouldn't have been surprised, but that got me banned as well.

I think that's not a way to treat people volunteering their time, so I'm no longer doing that.
I explained it on the ticket, back then.

**Pull request closed without merging**